net: lwm2m: Added Send, Composite Read/Write operation & SenML JSON support

jheiskan81 opened the pull request:
LwM2M SenML JSON support for Read & Write:
- LwM2M SenML JSON contain format support for Read / Write operation.
- Added Kconfig configurable for enable SenML JSON format.
- LwM2m read validate read operation and report out of memory.

LwM2M engine Send operation:
- Added Kconfig configurable for composite read path list size.
- Created generic Read object instance which is shared between general read and Composite Read operation.

LwM2M Composite-Read/Write operation support
- Composite Read operation support for CoAP FETCH method
- Composite Write operation support for CoAP iPATCH method

LwM2M engine default content format select

- LwM2M version 1.1 will select SenML Json for default content format.
- Version 1.0 will use TLV format.

SeppoTakalo committed:
net: lwm2m: Allow changing the protocol version to 1.1

This is a bare minimal changes to tell the server that we are using
LwM2M 1.1 version. Queue-mode parameter has changed between 1.0 and
1.1 so it must be changed in the same time.

Other 1.1 features may follow on separate commits. This is still
an experimental feature that allows developing and testing of
1.1 features.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

jheiskan81 commented:
Please review @SeppoTakalo , @jarlamsa @VeijoPesonen

SeppoTakalo commented on include/net/lwm2m.h:
```diff
@@ -1145,5 +1145,19 @@ void lwm2m_rd_client_update(void);
  */
 char *lwm2m_path_log_strdup(char *buf, struct lwm2m_obj_path *path);
 
+/* 
```
Missing the extra star.
You need `/**` to start a Doxygen block.

jheiskan81 replied:
OK I will fix this.

ww34ww replied:
How do I even engage in this zephyr conversation

SeppoTakalo commented on include/net/lwm2m.h:
```diff
@@ -1145,5 +1145,19 @@ void lwm2m_rd_client_update(void);
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, uint16_t content_format, char *path_list[], uint8_t path_list_size);
```
Why do we need content format as a parameter here?
I would assume that it is job for engine to choose which one to use from built in formats. You most probably will build in only one content format.

SeppoTakalo replied:
Can the `path_list` be `const char *`.
Also, can I free the list after this call? It should be documented how long I need to keep that in memory.

jheiskan81 replied:
Content format can be internally selected by engine I'm fine with that change. 

Path list need to at memory only time at function call. String path is converted to to internal used object path structure.

SeppoTakalo commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -370,6 +370,23 @@ config LWM2M_RW_JSON_SUPPORT
 
+
+config LWM2M_RW_SENML_JSON_SUPPORT
+	bool "support for SENML JSON writer[EXPERIMENTAL]"
```
What is the justification for this being EXPERIMENTAL?

The content format follows the standard, right?

jheiskan81 replied:
True I will remove that.

VeijoPesonen replied:
Experimental is justified in the sense that support for v1.1 is not fully in place. Experimental tells a user that you should really know what you're doing. But I guess it's only a matter of taste.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3217,6 +3261,131 @@ int lwm2m_perform_read_op(struct lwm2m_message *msg, uint16_t content_format)
+	/* Add object start mark */
+	engine_put_begin(&msg->out, &msg->path);
+
+	//Read resouse by path
```
"Read resource from path"

Also, use `/*  */` instead of `//` to follow a common style.

jheiskan81 replied:
Ok I will change to /* */ format.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1609 @@
+			ret = -EINVAL;
+			goto end_of_operation;
+		}
+		//Store Json File description flags
```
In this file, you have quite many `//` comments mixed with `/*  */` comment blocks.
Stick to one style, which I think should be `/* */`

jheiskan81 replied:
Ok I fix those.

SeppoTakalo reviewed (changes requested):
Left few comments.

jheiskan81 commented:
@SeppoTakalo I fix your comment's and also fix couple minor issue at Read operation.

rlubos commented:
Could you split unrelated chagnes to separate commits (SenML JSON formatter/send operation)? It not only results in a cleaner history, but also eases the review process, when you can see which change belongs to which feature.

jheiskan81 commented:
> Could you split unrelated chagnes to separate commits (SenML JSON formatter/send operation)? It not only results in a cleaner history, but also eases the review process, when you can see which change belongs to which feature.

OK I will split commit

Commit:
net: lwm2m: Added SenML JSON support for Read & Write

LwM2M SenML JSON contain format support for Read / Write operation.
Added Kconfig configurable for enable SenML JSON format.
LwM2m read validate read operation and report out of memory.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: lwm2m: Added LwM2M Send operation support

LwM2M engine Send operation support /dp.
Send operation based on Composite Read functionality
which is not enabled at engine side like composite write.
Added Kconfig configurable for composite read path list size.
Created generic Read object instance which is shared between
general read and Composite Read operation.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos I split original commit to 2 seperate. Please review again

jheiskan81 commented:
I fix CI warnings and add generic read object instance which is shared between Read and composite read operation.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1488 @@
+		if (block_ctx->base_name_stored) {
+			int base_name_length = 0;
+
+			if (block_ctx->base_name_path.level == 0) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/");
+			} else if (block_ctx->base_name_path.level == 1U) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/%u/",
+							    block_ctx->base_name_path.obj_id);
+			} else if (block_ctx->base_name_path.level == 2U) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/%u/%u/",
+							    block_ctx->base_name_path.obj_id,
+							    block_ctx->base_name_path.obj_inst_id);
+			} else if (block_ctx->base_name_path.level == 3U) {
+				base_name_length =
+					snprintk(base_name, sizeof(base_name), "/%u/%u/%u",
+						 block_ctx->base_name_path.obj_id,
+						 block_ctx->base_name_path.obj_inst_id,
+						 block_ctx->base_name_path.res_id);
+			} else {
+				base_name_length =
+					snprintk(base_name, sizeof(base_name), "/%u/%u/%u/%u",
+						 block_ctx->base_name_path.obj_id,
+						 block_ctx->base_name_path.obj_inst_id,
+						 block_ctx->base_name_path.res_id,
+						 block_ctx->base_name_path.res_inst_id);
+			}
```
Maybe it'd make sense to extract this as a utility function, and re-use at line 1319?

jheiskan81 replied:
yes good point I will do that. Only change is that path level need to be given by own parameter. Good spot

rlubos commented on include/net/lwm2m.h:
```diff
@@ -1145,5 +1145,18 @@ void lwm2m_rd_client_update(void);
+ * @param ctx LwM2M context
+ * @param path_list LwM2M Path string list
+ * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
+ *
```
One empty line shoud be enough.

jheiskan81 replied:
Ok

rlubos commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -378,6 +378,14 @@ config LWM2M_RW_SENML_JSON_SUPPORT
 	  Include support for write / parse SENML JSON data
 
+config LWM2M_COMPOSITE_PATH_LIST_SIZE
+	int "Maxinum # of composite read  and send operation url paths"
```
```suggestion
	int "Maximum # of composite read and send operation URL paths"
```

jheiskan81 replied:
Ok I will change like suggestion

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -1486,3 +1486,130 @@ int do_write_op_senml_json(struct lwm2m_message *msg)
+		LOG_ERR("No Valid Url at msg");
+		return -ESRCH;
+	}
+	int ret;
```
By convention we tend to keep the variable definitions together, at the beginning of the block.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5001,4 +5006,185 @@ static int lwm2m_engine_init(const struct device *dev)
+
+	struct lwm2m_obj_path path[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
+	int ret;
+
```
No need for this empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5001,4 +5006,185 @@ static int lwm2m_engine_init(const struct device *dev)
+		LOG_ERR("No space for End marker");
+		return -ENOMEM;
+	}
+
```
Redundant empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5001,4 +5006,185 @@ static int lwm2m_engine_init(const struct device *dev)
+
+}
+
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, char *path_list[], uint8_t path_list_size)
```
I wonder now, perhaps an application should have a chance to register result callback on it's own? So that the application knows whether the data was delivered or not. Up for discussion.

jheiskan81 replied:
Ok good point. Can we re-use some old callback function pointer?

rlubos replied:
I don't see any useful callback type already present for this purpose to be honest.

rlubos reviewed (commented):
Some initial feedback

jheiskan81 commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -370,6 +370,14 @@ config LWM2M_RW_JSON_SUPPORT
+config LWM2M_RW_SENML_JSON_SUPPORT
+	bool "support for SENML JSON contain format"
+	depends on BASE64
+	default n
```
Ok good to know.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1488 @@
+	return 0;
+}
+
+static size_t put_float32fix(struct lwm2m_output_context *out, struct lwm2m_obj_path *path,
+			     float32_value_t *value)
```
ok.

jheiskan81 replied:
get_float() and put_float() float are outdated I will fix those.

jheiskan81 commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -370,6 +370,14 @@ config LWM2M_RW_JSON_SUPPORT
 
+
+config LWM2M_RW_SENML_JSON_SUPPORT
+	bool "support for SENML JSON contain format"
```
how about: data format?

rlubos replied:
This also sounds fine

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1488 @@
+		}
+		return SENML_JSON_UNKNOWN_ATTRIBUTE;
+	}
+	if (buf_read(attrbute_name, fd->name_len, CPKT_BUF_READ(in->in_cpkt), &fd->name_offset) <
```
Ok

jheiskan81 commented:
@rlubos Thanks for good feedback. I update all things expect that lwm2m_engine_send() callback. I will investigate that later on.

jtguggedal commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4935,4 +5003,184 @@ static int lwm2m_engine_init(const struct device *dev)
+		goto cleanup;
+	}
+	LOG_INF("Send op to server (/dp)");
+	lwm2m_send_message(msg);
```
I'm a little curious about this call. It sort of conflicts with the rationale in this earlier commit reworking socket interaction: https://github.com/zephyrproject-rtos/zephyr/commit/32989a38f0c8911e83d64fc898e19d29b0800f75

Not saying it's wrong to do it like this, just want to ensure that we don't end up in a similar situation that triggered the previous rework.

jheiskan81 replied:
OK. I can change to asynch send. No problem. When I was testing feature I see I got message out quicker by using this API.

jtguggedal replied:
Yeah, it will get out faster this way. I was wondering if there's a way to get the same behavior (~instant send) using the async API, but that would require returning from `poll()` before it times out, which I don't know how to achieve.

jheiskan81 replied:
OK. I will change to asynch send call. Thanks for pointing this.

Commit:
net: lwm2m: LwM2M Composite read support

Adeed enabler for composite read to LwM2M engine.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: lwm2m2: LwM2M Composite-Write operation

CoAP method iPATCH enable LwM2M Composite-Write operation.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1654 @@
+
+			break;
+
+
```
Why there are so big gaps in the switch-case

jheiskan81 replied:
I will fix this.

VeijoPesonen commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -385,6 +385,21 @@ config LWM2M_RW_JSON_SUPPORT
+	default 6
+	help
+	  Define path list size for Composite Read and and send operation.
+	  Every entry in path list structure will increase stack usage by 13
```
Better not to put the exact number here. Just one more thing which might drift out of sync.

Commit:
net: lwm2m: Composite Observe operation placeholder

Added detetction for Composite Observervation request and cancel.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

VeijoPesonen reviewed (commented):
Good work! :)

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.h:
```diff
@@ -0,0 +1,26 @@
+/* General Write single Path operation */
+int do_write_op_senml_json(struct lwm2m_message *msg);
+
+/* Send opeartion builder */
```
Ah I will fix.

jheiskan81 replied:
I will fix this.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2279,68 +2300,70 @@ static int lwm2m_read_handler(struct lwm2m_engine_obj_inst *obj_inst,
 			return -ENOENT;
 		}
 
+		size_t read_put_data_length;
```
Yes I will waiting this PR and remove that part from the code and fix conflicts.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3824,45 +3990,35 @@ static int handle_request(struct coap_packet *request,
-					LOG_ERR("OBSERVE request missing token");
-					r = -EINVAL;
+					/* Composite Observation request & cancel handler */
+					//TODO add support for Composite observation support
```
Ok good point.

jheiskan81 commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -385,6 +385,21 @@ config LWM2M_RW_JSON_SUPPORT
+	default 6
+	help
+	  Define path list size for Composite Read and and send operation.
+	  Every entry in path list structure will increase stack usage by 13
```
Yes true I will remove that line.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3081,52 +3110,58 @@ static int do_read_op(struct lwm2m_message *msg, uint16_t content_format)
-	}
+	default:
+		LOG_ERR("Unsupported content-format: %u", content_format);
+		return -ENOMSG;
```
True I will change.

jheiskan81 replied:
I keep it ENOMSG like normal read is generated.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -137,6 +137,8 @@ BUILD_ASSERT(CONFIG_LWM2M_COAP_BLOCK_SIZE <= CONFIG_LWM2M_COAP_MAX_MSG_SIZE,
 /* buffer util macros */
 #define CPKT_BUF_WRITE(cpkt)	(cpkt)->data, &(cpkt)->offset, (cpkt)->max_len
 #define CPKT_BUF_READ(cpkt)	(cpkt)->data, (cpkt)->max_len
+#define CPKT_BUF_PTR(cpkt)	((cpkt)->data + (cpkt)->offset)
```
OK i will change to your proposal

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -137,6 +137,8 @@ BUILD_ASSERT(CONFIG_LWM2M_COAP_BLOCK_SIZE <= CONFIG_LWM2M_COAP_MAX_MSG_SIZE,
 #define CPKT_BUF_WRITE(cpkt)	(cpkt)->data, &(cpkt)->offset, (cpkt)->max_len
 #define CPKT_BUF_READ(cpkt)	(cpkt)->data, (cpkt)->max_len
+#define CPKT_BUF_PTR(cpkt)	((cpkt)->data + (cpkt)->offset)
+#define CPKT_BUF_SIZE(cpkt)	((cpkt)->max_len - (cpkt)->offset)
```
Ok

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3704,6 +3820,19 @@ static int handle_request(struct coap_packet *request,
+		msg->code = COAP_RESPONSE_CODE_CONTENT;
+		break;
+
+	case COAP_METHOD_IPATCH:
```
LwM2M v1.1 transport define success to 2.04

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3871,7 +4027,11 @@ static int handle_request(struct coap_packet *request,
 		case LWM2M_OP_WRITE:
 		case LWM2M_OP_CREATE:
-			r = do_write_op(msg, format);
+			if ((code & COAP_REQUEST_MASK) == COAP_METHOD_GET) {
```
OK good point I Will add comments.

jheiskan81 replied:
Ah there is bug also if ((code & COAP_REQUEST_MASK) == COAP_METHOD_IPATCH) { this should be for composite write

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -87,6 +94,13 @@ void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx);
 uint8_t *lwm2m_get_message_buf(void);
 int lwm2m_put_message_buf(uint8_t *buf);
 
+void lwm2m_engine_path_list_init(sys_slist_t *lwm2m_path_list, sys_slist_t *lwm2m_free_list,
```
OK. I will add

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4939,4 +5098,380 @@ static int lwm2m_engine_init(const struct device *dev)
+		return -ENOMEM;
+	}
+
+	msg->type = COAP_TYPE_CON;
```
Yes I will Add that boolean for that.

VeijoPesonen replied:
Maybe also combine it somehow with @rlubos's idea about having a callback for reporting status. An ACK doesn't do too much good without that. Now it seems to be used for tracing purposes.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4939,4 +5098,380 @@ static int lwm2m_engine_init(const struct device *dev)
+	/* Write requested path data */
+	ret = do_send_op(msg, content_format, &lwm_path_list);
+	if (ret < 0) {
+		LOG_ERR("error in multi-format read (err:%d)", ret);
```
That is fixed

Commit:
net: lwm2m: LwM2M engine default content format select

LwM2M version 1.1 will select SenML Json for default content format.
Version 1.0 will use TLV format.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos I have been rebased PR and update Send API for use const char pointer type.

jheiskan81 commented:
@SeppoTakalo Please review this.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -87,6 +94,16 @@ void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx);
+/* Helper function for Add new Path to list */
+int lwm2m_engine_add_path_to_list(sys_slist_t *lwm2m_path_list, sys_slist_t *lwm2m_free_list,
+				  struct lwm2m_obj_path *path);
+/* Helper function for Clear Path which are included to shorter path already */
```
```suggestion
/* Remove paths when parent already exist in the list. */
```

jheiskan81 replied:
ok

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -87,6 +94,16 @@ void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx);
+void lwm2m_engine_path_list_init(sys_slist_t *lwm2m_path_list, sys_slist_t *lwm2m_free_list,
+				 struct lwm2m_obj_path_list path_object_buf[],
+				 uint8_t path_object_size);
+/* Helper function for Add new Path to list */
```
```suggestion
/* Add new Path to the list */
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -87,6 +94,16 @@ void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx);
 uint8_t *lwm2m_get_message_buf(void);
 int lwm2m_put_message_buf(uint8_t *buf);
 
+/* Helper function for Init object path list */
```
```suggestion
/* Initialize path list */
```

jheiskan81 replied:
ok

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3745,22 +3882,28 @@ static int handle_request(struct coap_packet *request,
-			/* No matching object found - ignore request */
-			r = -ENOENT;
-			goto error;
+	/* Do Only Object find if path have been parsed */
```
```suggestion
	/* Search for the object only if the path have been parsed. */
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5020,4 +5172,383 @@ static int lwm2m_engine_init(const struct device *dev)
+	bool add_before_current = false;
+
+	if (path->level == 0) {
+		/* Clear List if some one add root level */
```
```suggestion
		/* Clear the list if we are adding the root path which includes all */
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5020,4 +5172,383 @@ static int lwm2m_engine_init(const struct device *dev)
+		lwm2m_engine_free_list(lwm2m_path_list, lwm2m_free_list);
+	}
+
+	/* Check is it at list already here */
```
```suggestion
	/* Check if there is room to add more */
```

jheiskan81 replied:
ok

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3745,22 +3882,28 @@ static int handle_request(struct coap_packet *request,
 	} else {
-		LOG_DBG("No accept option given. Assume OMA TLV.");
-		accept = LWM2M_FORMAT_OMA_TLV;
+		/* Select Default based LWM2M_VERSION */
```
ok

jheiskan81 commented:
@SeppoTakalo Thanks for feedback I have been fixed those raised point.

jheiskan81 commented:
I update Composite Read to support root path read. Now it pass lwm2m v1.1 INT-235 conformance test when Device, Server and Portfolio object are registered and CONFIG_LWM2M_COAP_MAX_MSG_SIZE=1024.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1206,6 +1218,13 @@ static int select_reader(struct lwm2m_input_context *in, uint16_t format)
 		break;
 #endif
 
+
```
Redundant empty line.

jheiskan81 replied:
Ah I will fix that.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3136,6 +3157,12 @@ static int do_read_op(struct lwm2m_message *msg, uint16_t content_format)
 		return do_read_op_json(msg, content_format);
 #endif
 
+
```
Redundant empty line.

jheiskan81 replied:
I will fix this.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5020,4 +5052,185 @@ static int lwm2m_engine_init(const struct device *dev)
+	struct lwm2m_obj_path_list *new_entry;
+	bool add_before_current = false;
+
+	if (path->level == 0) {
```
General note: would be nice to use symbols for PATH levels instead of numbers.

jheiskan81 replied:
I will update my changes to start using thosen defined constant.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1554 @@
+
+	if (fd->base_name.level == 0) {
+		return;
+	}
```
Alignment seems off here.

jheiskan81 replied:
Ok.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1554 @@
+
+	SYS_SLIST_FOR_EACH_CONTAINER(lwm_path_list, entry, node) {
+		if (fd->base_name_used == false) {
+	/* First at list is define compare for rest */
```
Alignment seems off here.

jheiskan81 replied:
ok

rlubos commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -392,6 +392,12 @@ config LWM2M_RW_SENML_JSON_SUPPORT
 	  Include support for write / parse SENML JSON data
 
+config LWM2M_COMPOSITE_PATH_LIST_SIZE
+	int "Maximum # of composite read  and send operation URL paths"
```
```suggestion
	int "Maximum # of composite read and send operation URL paths"
```

jheiskan81 replied:
OK. I fix this.

rlubos commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -392,6 +392,12 @@ config LWM2M_RW_SENML_JSON_SUPPORT
+	int "Maximum # of composite read  and send operation URL paths"
+	default 6
+	help
+	  Define path list size for Composite Read and and send operation.
```
Double "and"

jheiskan81 replied:
ok.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5233,4 +5238,242 @@ void lwm2m_engine_clear_duplicate_path(sys_slist_t *lwm2m_path_list, sys_slist_t
+		if (ret == -ENOMEM) {
+			return ret;
+		}
+
+
```
No need for this gap here,

jheiskan81 replied:
Ok

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3765,6 +3765,24 @@ static int lwm2m_engine_observation_handler(struct lwm2m_message *msg, int obser
 
+static int lwm2m_engine_default_content_format(uint16_t *accept_format)
+{
+#if LWM2M_PROTOCOL_VERSION_MAJOR == 1 && LWM2M_PROTOCOL_VERSION_MINOR == 1
```
I think it'd be cleaner to wait for #42169 and use `IS_ENABLED(CONFIG_LWM2M_VERSION_1_1)`

jheiskan81 replied:
OK I Update that Part. It not pass a compile process untill #42169 is merged.

tejlmand commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -385,6 +385,13 @@ config LWM2M_RW_JSON_SUPPORT
 
+
+config LWM2M_RW_SENML_JSON_SUPPORT
+	bool "support for SENML JSON data format"
```
Minor nit, as per the guidelines, consider to just write: `SENML JSON data format`:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

```suggestion
	bool "SENML JSON data format"
```

jheiskan81 replied:
OK I will fix that.

tejlmand commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:
```diff
@@ -0,0 +1,1554 @@
+					fd->json_flags &= ~T_STRING_BEGIN;
+					fd->json_flags |= T_STRING_END;
+				} else {
+					fd->json_flags &= ~T_STRING_END;
+					fd->json_flags |= T_STRING_BEGIN;
+				}
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		case ':':
+			if (!escape) {
+				fd->json_flags &= ~T_STRING_END;
+				fd->json_flags |= T_VALUE;
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		/* ignore whitespace */
+		case ' ':
+		case '\n':
+		case '\t':
+			if (!(fd->json_flags & T_STRING_BEGIN)) {
+				break;
+			}
+
+			__fallthrough;
+
+		default:
+			json_add_char(in, fd);
+		}
+
+		if (escape) {
+			escape = false;
+		}
+	}
```
This code is almost identical to this:
https://github.com/zephyrproject-rtos/zephyr/blob/ccf51e3cd3ae455d94fbf3706e57c62cc3fc318b/subsys/net/lib/lwm2m/lwm2m_rw_json.c#L154-L248

and the same is probably true for other parts of this file.

Furthermore we have a general json library here, for parsing of json:
https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/json.c

Controlled with [CONFIG_JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html)

Why do we need so many versions / copies of json parsers ?
It increases maintenance burden, not to mention, if one library fixes a bug, that very same bug has a risk of being present also in another lib.

Could you please align to a single implementation for json.

jheiskan81 replied:
I looking for option to use CONFIG_JSON_LIBRARY. I need to study what need I need to change if I start looking for. I propose that I integrated SenML Json part first to use CONFIG_JSON_LIBRARY.

tejlmand replied:
Followup task created: https://github.com/zephyrproject-rtos/zephyr/issues/43103

tejlmand commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3172,6 +3172,23 @@ static int do_read_op(struct lwm2m_message *msg, uint16_t content_format)
 	}
 }
 
+
```
double empty line
```suggestion
```

jheiskan81 replied:
Ok I will that.

tejlmand commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3612,6 +3612,24 @@ static int do_write_op(struct lwm2m_message *msg,
+	}
+}
+
+
+
```
multiple empty lines.

```suggestion
```

tejlmand commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3728,6 +3728,43 @@ static bool lwm2m_engine_path_included(uint8_t code, bool bootstrap_mode)
+		if (r < 0) {
+			LOG_ERR("add OBSERVE error: %d", r);
+		}
+
```
could you please be consistent in whether you have a empty line or not right before the `} else ... {`  .
Right now you mix like this:
```
if (...) {
    {
    }
<empty line>
} else if (...) {
    {
    }
} else {
   ...
}
```

```suggestion
```

jheiskan81 replied:
Thanks for good point I will clean those empty lines.

tejlmand reviewed (changes requested):
Please make existing lwm2m json code re-usable or use the [JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html) for json parsing.

We should not have three json implementations doing the same in the Zephyr code base.

jheiskan81 commented:
> Please make existing lwm2m json code re-usable or use the [JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html) for json parsing.
> 
> We should not have three json implementations doing the same in the Zephyr code base.

Ah good point. that will need a extra job.
I was implementing SenML Json it based on old lwm2m_rw_json.c which was not using json library.
I need to study how I can update SenML file to use parser and decode part by using json library.

jheiskan81 commented:
@rlubos and @tejlmand I have been fix minor changes. I´m still pending JSON Library update. I would propose that we do that separately at own PR.

SeppoTakalo commented:
> Please make existing lwm2m json code re-usable or use the [JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html) for json parsing.
> 
> We should not have three json implementations doing the same in the Zephyr code base.

@tejlmand What you are proposing, makes sense for the long term.

However, the code submitted now, follows the way how the library have been earlier, so it should not be a reason to block PRs. We have timelines to catch.

For the future, we have recorded this request into our backlog, and once we have time, we are happy to do it. But in the mean time, please remove the "Request for changes" marker, as we don't have time to do it in the first round.

tejlmand commented:
> For the future, we have recorded this request into our backlog, and once we have time, we are happy to do it. 

As discussed offline.
Please open an enhancement issue upstream here: https://github.com/zephyrproject-rtos/zephyr/issues/new/choose and link to this issue in this PR.
The community doesn't have access to your internal backlog.

Also please provide a rough timeframe for the followup work, for example is this expected to be completed before 3.1 ?

tejlmand reviewed (approved):
lgtm, Followup task created: https://github.com/zephyrproject-rtos/zephyr/issues/43103

rlubos reviewed (approved):
LGTM, please rebase.

SeppoTakalo commented:
As proposed by @tejlmand we are aiming to improve this before Zephyr 3.1 release which I believe is scheduled somewhere in beginning of July.

SeppoTakalo commented:
Github seem to be broken somehow.. This branch should be based on main as can be seen here https://github.com/jheiskan81/zephyr/tree/lwm2m_send_senml_json

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -23,6 +27,7 @@
 #define LWM2M_FORMAT_APP_OCTET_STREAM	42
 #define LWM2M_FORMAT_APP_EXI		47
 #define LWM2M_FORMAT_APP_JSON		50
+#define LWM2M_FORMAT_APP_SEML_JSON	110
```
Typo, should be _LWM2M_FORMAT_APP_SENML_JSON_.
